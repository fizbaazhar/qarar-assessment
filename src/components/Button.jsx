import React from 'react';

/**
 * Button Component
 * Reusable button with different variants, sizes, and states
 * 
 * @param {string} variant - 'primary', 'secondary', 'danger'
 * @param {string} size - 'sm', 'md', 'lg'
 * @param {boolean} loading - Shows loading spinner
 * @param {boolean} disabled - Disables the button
 * @param {string} className - Additional CSS classes
 * @param {React.ReactNode} children - Button content
 * @param {Object} props - All other button props
 */
const Button = ({ 
  variant = 'primary', 
  size = 'md', 
  loading = false, 
  disabled = false, 
  className = '', 
  children, 
  ...props 
}) => {
  // Base button classes
  const baseClasses = 'font-bold rounded-lg focus:outline-none focus:ring-2 focus:ring-offset-2 transition-all duration-200 disabled:opacity-50 disabled:cursor-not-allowed';
  
  // Variant classes
  const variantClasses = {
    primary: 'bg-primary text-white border-2 border-transparent hover:bg-white hover:text-primary hover:border-primary focus:ring-primary',
    secondary: 'bg-gray-200 hover:bg-gray-300 text-gray-800 focus:ring-gray-500',
    danger: 'bg-red-500 hover:bg-red-600 text-white focus:ring-red-500',
  };
  
  // Size classes
  const sizeClasses = {
    sm: 'px-3 py-1.5 text-sm',
    md: 'px-4 py-2 text-sm',
    lg: 'px-6 py-3 text-base',
  };
  
  // Combine all classes
  const buttonClasses = `${baseClasses} ${variantClasses[variant]} ${sizeClasses[size]} ${className}`;
  
  return (
    <button 
      className={buttonClasses}
      disabled={disabled || loading}
      {...props}
    >
      {loading ? (
        <div className="flex items-center justify-center">
          <div className="animate-spin rounded-full h-5 w-5 border-b-2 border-current mr-2"></div>
          Loading...
        </div>
      ) : (
        children
      )}
    </button>
  );
};

export default Button; 