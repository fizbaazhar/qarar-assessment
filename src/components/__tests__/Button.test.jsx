import { describe, it, expect, vi } from 'vitest'
import { render, screen, fireEvent } from '../../test/test-utils'
import Button from '../Button'

describe('Button Component', () => {
  // Test basic rendering
  it('renders button with children text', () => {
    render(<Button>Click me</Button>)
    expect(screen.getByRole('button', { name: 'Click me' })).toBeInTheDocument()
  })

  // Test different variants
  it('applies primary variant styles correctly', () => {
    render(<Button variant="primary">Primary Button</Button>)
    const button = screen.getByRole('button', { name: 'Primary Button' })
    expect(button).toHaveClass('bg-primary', 'text-white')
  })

  it('applies secondary variant styles correctly', () => {
    render(<Button variant="secondary">Secondary Button</Button>)
    const button = screen.getByRole('button', { name: 'Secondary Button' })
    expect(button).toHaveClass('bg-gray-200', 'text-gray-800')
  })

  it('applies danger variant styles correctly', () => {
    render(<Button variant="danger">Danger Button</Button>)
    const button = screen.getByRole('button', { name: 'Danger Button' })
    expect(button).toHaveClass('bg-red-500', 'text-white')
  })

  // Test different sizes
  it('applies small size styles correctly', () => {
    render(<Button size="sm">Small Button</Button>)
    const button = screen.getByRole('button', { name: 'Small Button' })
    expect(button).toHaveClass('px-2.5', 'text-[12px]')
  })

  it('applies medium size styles correctly', () => {
    render(<Button size="md">Medium Button</Button>)
    const button = screen.getByRole('button', { name: 'Medium Button' })
    expect(button).toHaveClass('px-3', 'text-xs')
  })

  it('applies large size styles correctly', () => {
    render(<Button size="lg">Large Button</Button>)
    const button = screen.getByRole('button', { name: 'Large Button' })
    expect(button).toHaveClass('px-4', 'text-xs')
  })

  // Test loading state
  it('shows loading spinner when loading is true', () => {
    render(<Button loading={true}>Loading Button</Button>)
    expect(screen.getByText('Loading...')).toBeInTheDocument()
    expect(screen.getByRole('button')).toBeDisabled()
  })

  // Test disabled state
  it('disables button when disabled is true', () => {
    render(<Button disabled={true}>Disabled Button</Button>)
    const button = screen.getByRole('button', { name: 'Disabled Button' })
    expect(button).toBeDisabled()
    expect(button).toHaveClass('disabled:opacity-50', 'disabled:cursor-not-allowed')
  })

  // Test click handler
  it('calls onClick handler when clicked', () => {
    const handleClick = vi.fn()
    render(<Button onClick={handleClick}>Clickable Button</Button>)
    
    const button = screen.getByRole('button', { name: 'Clickable Button' })
    fireEvent.click(button)
    
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  // Test that disabled button does not call onClick
  it('does not call onClick when disabled', () => {
    const handleClick = vi.fn()
    render(<Button disabled={true} onClick={handleClick}>Disabled Button</Button>)
    
    const button = screen.getByRole('button', { name: 'Disabled Button' })
    fireEvent.click(button)
    
    expect(handleClick).not.toHaveBeenCalled()
  })

  // Test custom className
  it('applies custom className', () => {
    render(<Button className="custom-class">Custom Button</Button>)
    const button = screen.getByRole('button', { name: 'Custom Button' })
    expect(button).toHaveClass('custom-class')
  })

  // Test button type
  it('applies correct button type', () => {
    render(<Button type="submit">Submit Button</Button>)
    const button = screen.getByRole('button', { name: 'Submit Button' })
    expect(button).toHaveAttribute('type', 'submit')
  })

  // Test accessibility
  it('supports aria-label for accessibility', () => {
    render(<Button aria-label="Custom label">Button</Button>)
    const button = screen.getByRole('button', { name: 'Custom label' })
    expect(button).toBeInTheDocument()
  })

  // Test snapshot
  it('matches snapshot for primary button', () => {
    const { container } = render(<Button variant="primary">Snapshot Button</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })
}) 