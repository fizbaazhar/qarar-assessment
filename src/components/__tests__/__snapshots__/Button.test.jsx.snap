// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > matches snapshot for primary button 1`] = `
<button
  class="font-bold rounded-lg focus:outline-none focus:ring-2 focus:ring-offset-2 transition-all duration-200 disabled:opacity-50 disabled:cursor-not-allowed flex items-center justify-center gap-2 whitespace-nowrap bg-primary text-white border-2 border-transparent hover:bg-white hover:text-primary hover:border-primary focus:ring-primary active:scale-95 px-3 sm:px-4 py-1.5 sm:py-1.5 text-xs sm:text-base "
>
  Snapshot Button
</button>
`;
